### PR TITLE
chore: Add pytest and selenium test framework to the project

### PR DESCRIPTION
**Pull request by m1yag1**
Added pytest and selenium test framework to the project

### CODE REVIEW

File: .gitignore
Comment (TomWoodward) [inline .gitignore]: i'm not morally opposed to ignoring editor files in the project gitignore, but since your editor is a personal decision, you could add them to your [global gitignore](https://help.github.com/articles/ignoring-files/#create-a-global-gitignore)

File: .gitignore
Comment (TomWoodward) [inline .gitignore]: `.env` is intentionally allowed, use `.env.local` for stuff outside version control. that said we don't use the environment files for the JS on this project anymore so if you want to leave this you can. if you do leave it maybe combine it with all the env related stuff above

File: .gitignore
Comment (TomWoodward) [inline .gitignore]: i feel as if some of this stuff isn't necessary

File: .pre-commit-config.yaml
Comment (TomWoodward) [inline .pre-commit-config.yaml]: @philschatz maybe we should add the faster checks like `tsc` and `tslint` as commit hooks?

File: Makefile-py
Comment (TomWoodward) [inline Makefile-py]: can this be in the folder with all the python stuff? or rename it to just `Makefile` and i'll start adding rando scripts here too

File: pyproject.toml
Comment (TomWoodward) [inline pyproject.toml]: if we could move as many of the python project files into a subdirectory as possible that would make me happy

File: testrail.example.cfg
Comment (TomWoodward) [inline testrail.example.cfg]: i know this is just an example, but in reality we're going to have to eventually pull all these configs from the environment, is that possible?

File: tests/ui/README.md
Comment (TomWoodward) [inline tests/ui/README.md]: rex-web

**Review by TomWoodward — COMMENTED** [review]
what is there so far of the page structure and framework looks really good. i think all my concerns were about merging the py and ts project configs at the top level of the repo. it would make me feel better to move the python specific configs into a subdirectory

File: .gitignore
Comment (m1yag1) [reply .gitignore]: I can put it in my global gitignore. no biggie.

File: .gitignore
Comment (m1yag1) [reply .gitignore]: Yea, again, it's the cause of some copy pasta.. I'll remove it.

File: pyproject.toml
Comment (m1yag1) [reply pyproject.toml]: Sure, I explored this at first but there were a few things that looked to the root directory for configs (the big one is black which is the code formatter for python) and the work arounds were like, "create a bash script that automatically adds the --config-file /some/path/to/config". My main idea was to go for easy to run w/o having someone remember a bunch of command line flags for config files.

Additionally, I've worked in a number of projects that use python and js. I don't think I've ever been like, "let me shove the node config stuff into another folder b/c it's not python." The root allows one to see the files necessary to run the project. However, I do understand the need to keep the root clean and only have files necessary for the project's runtime. An example of python and js living happily together is here: https://github.com/openstax/research-eGrader

I'll try to get things working where python is hidden. That may require more time investment than I have this week so I may not finish this soon as next week I have other things I need to work on.

File: testrail.example.cfg
Comment (m1yag1) [reply testrail.example.cfg]: Hmm.. I don't think the pytest-testrail plugin uses env vars. I'll have to double check.

File: pyproject.toml
Comment (m1yag1) [reply pyproject.toml]: oh hypothesis is another example that has js stuff in the root along with the python: https://github.com/hypothesis/h

File: .gitignore
Comment (m1yag1) [reply .gitignore]: Also, .DS_Store is in there.. i had added it again on accident, but that would also apply to your comment too. That's a OSX specific thing and operating systems to develop are a personal choice too ;).

File: .gitignore
Comment (m1yag1) [reply .gitignore]: Ah, that's a good practice. I'll have to start doing that with the other projects we have. 👍

File: .pre-commit-config.yaml
Comment (m1yag1) [reply .pre-commit-config.yaml]: I really like the use of pre-commit hooks. I think they are a great way to run the checks as you commit instead of waiting for some other time. Even though I've had linters run based on the test command I always seem to break CI for some dumb linting error. Now that I use pre-commit hooks it doesn't happen anymore.

File: Makefile-py
Comment (m1yag1) [reply Makefile-py]: My thought was that if other Make commands were introduced you could create one Make file that combines them all. It can be useful to separate out the docker specific Make stuff using something like `make -f Makefile-docker` but also combine them with a core Makefile and use `include Makefile-docker`. I thought the separation might help but maybe not. 🤷‍♂️

File: testrail.example.cfg
Comment (m1yag1) [reply testrail.example.cfg]: Yea, pytest-rail out of the box doesn't support environment variables. I'll need to add those.

File: Makefile-py
Comment (TomWoodward) [reply Makefile-py]: having sub-makefiles and composing them sounds like a cool pattern, in that case could we move the Makefile-py into a python subdirectory? or would that mess with the cwd of the scripts and be confusing? i'm not _opposed_ to having stuff in the top level, i'd just like to put things in subdirectories that can be.

File: pyproject.toml
Comment (TomWoodward) [reply pyproject.toml]: yea, again, not opposed. the idea is that we would be working in the py tests regularly (for almost every feature developed probably) and i'm not trying to make the py code second class, even for the ts files i push them down the tree as much as i can

File: testrail.example.cfg
Comment (TomWoodward) [reply testrail.example.cfg]: ok, in CI we can specify them via command line, that'll be easier

Comment (m1yag1) [reply testrail.example.cfg]: 👍

File: Makefile-py
Comment (m1yag1) [reply Makefile-py]: yes we could move the py Makefile to a subdirectory.

File: tests/ui/test_toc.py
Comment (m1yag1) [inline tests/ui/test_toc.py]: @TomWoodward Here I added a basic test of clicking the TOC toggle button. This test currently passes for both desktop and mobile resolutions. However, in the mobile view it really covers up the screen so I would suggest that the mobile table of contents be hidden by default.

File: Makefile-py
Comment (m1yag1) [reply Makefile-py]: I moved it. This is resolved.

File: pyproject.toml
Comment (m1yag1) [reply pyproject.toml]: I was able to move most of the files into the tests directory. The only one I wasn't able to is the config for pre-commit.

File: testrail.example.cfg
Comment (m1yag1) [reply testrail.example.cfg]: I removed the file and added instructions in the readme to create a file if someone wanted to run it locally and create their own run.

File: tests/ui/README.md
Comment (m1yag1) [reply tests/ui/README.md]: Fixed!

File: tests/ui/test_toc.py
Comment (TomWoodward) [reply tests/ui/test_toc.py]: i believe this is the plan you can see the upcoming new designs here openstax/unified-issues#104

File: tests/ui/pages/base.py
Comment (TomWoodward) [reply tests/ui/pages/base.py]: in javascript i would create a util decorator function that did the size guard and sleep that you just pass a function reference for `set_window_size` or `set_window_position` into.

i'm not sure if there is a comparably easy thing to do in python, but these two blocks of code are nearly identical.

File: tests/ui/pages/base.py
Comment (TomWoodward) [inline tests/ui/pages/base.py]: is there no way to check the window size to verify this worked instead of arbitrarily sleeping?

File: tests/ui/pages/content.py
Comment (TomWoodward) [inline tests/ui/pages/content.py]: @philschatz should weigh in but i think we should just sprinkle some `aria-label`s on these to select them by

File: tests/ui/pages/content.py
Comment (TomWoodward) [inline tests/ui/pages/content.py]: we should really use the js hook that the app provides, that is the canonical way that we're identifying when the page is done doing _stuff_

File: tests/ui/regions/base.py
Comment (TomWoodward) [reply tests/ui/regions/base.py]: similar to the other situation, i'd consider changing this to `return self.page.scroll_to(element or self.root)` or `return self.page.scroll_to(defaultElement(element))`

**Review by TomWoodward — CHANGES_REQUESTED** [review]

File: tests/ui/pages/content.py
Comment (m1yag1) [reply tests/ui/pages/content.py]: I agree. They didn't exist so this is what i did to get the tests to identify the right elements.

File: tests/ui/pages/content.py
Comment (TomWoodward) [reply tests/ui/pages/content.py]: when we run these tests against a built environment the page will be complete before the js is even loaded, so then this code would thought the page was done loading but all interactions would fail

File: tests/ui/pages/base.py
Comment (m1yag1) [reply tests/ui/pages/base.py]: I removed this as it's probably better to implement when we need it.

File: tests/ui/pages/base.py
Comment (m1yag1) [reply tests/ui/pages/base.py]: I removed this code like the previous code. This was done for some reason like it b/c of something with cnx-automation. we probably don't need to bring that code over unless the specifics of the test are similar.

File: tests/ui/pages/content.py
Comment (m1yag1) [reply tests/ui/pages/content.py]: Could you provide me with some js code that would allow me to resolve the js hook. It would look something like this.

https://github.com/openstax/cnx-automation/blob/try-applitools/tests/webview/ui/test_applitools.py#L25-L27

I have no idea what the return on something like that is.. i would need to experiment.

**Comment by TomWoodward** [comment]
@m1yag1 when i try to run `pytest --base-url http://localhost:3000 tests/ui` i get 

```bash
============================= test session starts ==============================
platform darwin -- Python 3.7.0, pytest-4.1.1, py-1.7.0, pluggy-0.8.1 -- /Users/tomwoodward/code/openstax/books-web/.venv/bin/python3
cachedir: .pytest_cache
sensitiveurl: ^(?:https?\:\/\/)?openstax\.org
metadata: {'Python': '3.7.0', 'Platform': 'Darwin-17.7.0-x86_64-i386-64bit', 'Packages': {'pytest': '4.1.1', 'py': '1.7.0', 'pluggy': '0.8.1'}, 'Plugins': {'variables': '1.7.1', 'selenium': '1.15.1', 'metadata': '1.8.0', 'html': '1.20.0', 'base-url': '1.4.1'}, 'Base URL': '', 'Driver': None, 'Capabilities': {}}
rootdir: /Users/tomwoodward/code/openstax/books-web/tests/ui, inifile: setup.cfg
plugins: variables-1.7.1, selenium-1.15.1, metadata-1.8.0, html-1.20.0, base-url-1.4.1
collecting ... collected 0 items / 1 errors

==================================== ERRORS ====================================
_________________________ ERROR collecting test_toc.py _________________________
ImportError while importing test module '/Users/tomwoodward/code/openstax/books-web/tests/ui/test_toc.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
tests/ui/test_toc.py:1: in <module>
    from tests.ui.pages.content import Content
E   ModuleNotFoundError: No module named 'tests'
- generated html file: /Users/tomwoodward/code/openstax/books-web/ui-test.html -
!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!!
=========================== 1 error in 0.17 seconds ============================
```

i haven't installed the webdriver stuff yet, but this doesn't look related i don't think...

File: README.md
Comment (TomWoodward) [inline README.md]: all of our other tests are distributed in the js code next to the stuff that they're testing. we have a /src/test place for making testing utilities, mocks, etc.

because of this, i don't think we're ever going to want to put anything else in the top level /tests folder, and having it wrap the single ui folder seems a little weird. can we rename the whole thing to /selenium or /ui-tests or something like that?

**Review by TomWoodward — CHANGES_REQUESTED** [review]

File: README.md
Comment (m1yag1) [reply README.md]: I figured this would come up and just picked tests as that was standard for python. I also wasn't sure if we would employ python to do some other kind of testing like integration tests... so i split it up like that... thinking it may be `tests(or something else)/ui`,`tests/integration`, `tests/visual`

maybe just name the root folder to `pytest`?

Comment (m1yag1) [reply README.md]: I based it off what mozilla had. https://github.com/mozilla/addons-server/tree/master/tests

Comment (TomWoodward) [reply README.md]: 👍 for root folder `pytest`, if you think its prudent to make a subfolder below that go for it

File: Makefile
Comment (TomWoodward) [inline Makefile]: this line is also in the py makefile, does it need to have the same commands in both places?

File: pytest-selenium/README.md
Comment (TomWoodward) [inline pytest-selenium/README.md]: this doesn't exist

File: pytest-selenium/tests/conftest.py
Comment (TomWoodward) [inline pytest-selenium/tests/conftest.py]: resolutions duplicated here and above?

**Review by TomWoodward — CHANGES_REQUESTED** [review]

File: Makefile
Comment (m1yag1) [reply Makefile]: i believe so. if someone were to do `make -f pytest-selenium/Makefile-py`

Comment (m1yag1) [reply Makefile]: However, that is not likely so i'll remove it ;)

File: pytest-selenium/tests/conftest.py
Comment (m1yag1) [reply pytest-selenium/tests/conftest.py]: The ids are mainly for display when the tests are running. It's not actually duplicating them but rather adding a label.

Comment (m1yag1) [reply pytest-selenium/tests/conftest.py]: Example:
image

**Review by TomWoodward — CHANGES_REQUESTED** [review]

File: pytest-selenium/tests/conftest.py
Comment (TomWoodward) [reply pytest-selenium/tests/conftest.py]: but why not use the constants in these strings?

File: README.md
Comment (m1yag1) [reply README.md]: Haha, yea you're right

File: pytest-selenium/README.md
Comment (m1yag1) [reply pytest-selenium/README.md]: Yes, a lot of these have changed with the new changes. I'll correct them.

File: pytest-selenium/tests/conftest.py
Comment (m1yag1) [reply pytest-selenium/tests/conftest.py]: I did

File: README.md
Comment (m1yag1) [reply README.md]: yea sorry.. I typically pass in Chrome as a default. Since we have a focus on cross browser it's something more explicit. I'm not used to it having to be passed in manually.

**Review by TomWoodward — DISMISSED** [review]
this is working both locally and through saucelabs and can post the results back to testrail. i'll make some additional changes when i try to automate things more (the `make test` command is mostly just note taking for me) but i think we should get this merged down! :shipit:

**Review by philschatz — CHANGES_REQUESTED** [review]
This is just a review of the precommit hooks, `Makefile`, and `README.md` (awesome to read through!) but will submit another one for the actual code

File: pytest-selenium/pages/content.py
Comment (philschatz) [inline pytest-selenium/pages/content.py]:
```suggestion
    _table_of_contents_button_locator = (By.CSS_SELECTOR, "[aria-label='Click to close the Table of Contents']")
```

Relying on the `[aria-label]` rather than a fragile selector helps test accessibility _and_ is less fragile than a `>` selector

File: pytest-selenium/pages/content.py
Comment (philschatz) [inline pytest-selenium/pages/content.py]: What are some reasons for using very-specific selectors in the Selenium tests?

```suggestion
        'h1',
```

If the WCAG guidelines are followed (I think we do so far) there should _only_ be one `h1` in the whole document. And if we don't follow the guidelines then something should blow up anyway 😄

File: pytest-selenium/pages/content.py
Comment (philschatz) [inline pytest-selenium/pages/content.py]: Does `self.find_element` throw an error when the locator matches `0` or `> 1` elements? If not, would it make sense to add a help that does that?

File: pytest-selenium/pages/content.py
Comment (philschatz) [inline pytest-selenium/pages/content.py]:
```suggestion
    _table_of_contents_nav_locator = (By.CSS_SELECTOR, "nav")
```

File: .pre-commit-config.yaml
Comment (TomWoodward) [reply .pre-commit-config.yaml]: i would definitely not want a pre-commit hook to run tests because i don't want to wait for them every time i commit. the linter is pretty fast though. i like to have control over when my code gets re-formatted and would not love to have it happen automatically on commit

File: pytest-selenium/pages/content.py
Comment (philschatz) [inline pytest-selenium/pages/content.py]: Is this different than the `_table_of_contents_nav_locator` ? If not, can they be combined?

File: pytest-selenium/setup.cfg
Comment (philschatz) [inline pytest-selenium/setup.cfg]: What is this field used for? I don't see it used anywhere directly... maybe it is used by an included python package?

File: pytest-selenium/pages/base.py
Comment (TomWoodward) [reply pytest-selenium/pages/base.py]: the reason we're using python in the first place is because mike has already solved this problem using python and doesn't have to re-invent everything to set up rex in the same way. down the line once we get things working i'd consider porting the suite over to typescript. in the meantime i don't think its worth going too far out of our way to modify the base pytest framework (like adding types)

File: pytest-selenium/tests/conftest.py
Comment (philschatz) [inline pytest-selenium/tests/conftest.py]: Is there a reason these are defined as `height,width` here but they are used in both the code and the docs as `width,height`?

File: pytest-selenium/tests/conftest.py
Comment (philschatz) [inline pytest-selenium/tests/conftest.py]: Could these be conditional on running in a container (environment `CI=true` or `TRAVIS=true`)? Once we start storing cookies (logged-in, whether popups were dismissed, etc) it might cause tests to fail unexpectedly.

File: pytest-selenium/tests/test_toc.py
Comment (philschatz) [inline pytest-selenium/tests/test_toc.py]: ❤️ The whole test is so sooo concise and readable 😄

File: README.md
Comment (philschatz) [inline README.md]:
```suggestion
make venv
source ./.venv/bin/activate
run these with `pytest --base-url http://localhost:3000 --driver Chrome pytest-selenium/tests`
```

This _partially_ worked for me but I got the following error (do we need separate instructions for installing chromedriver?):

```
selenium.common.exceptions.WebDriverException: Message: 'chromedriver' executable needs to be in PATH.
```

Maybe it would make sense to add a Makefile target instead? Like `make test-local` or something?

**Review by philschatz — CHANGES_REQUESTED** [review]

File: .pre-commit-config.yaml
Comment (m1yag1) [reply .pre-commit-config.yaml]: yea, i don't run tests.. i don't like that. only linting checks on the code i'm committing. We can just use this for the python stuff. I can move it to the python Makefile.

File: pytest-selenium/README.md
Comment (m1yag1) [reply pytest-selenium/README.md]: Most of the time i believe this will be ran by automation. These instructions are here in the event if someone wants to do that locally which may not be that often.

File: pytest-selenium/pages/base.py
Comment (m1yag1) [reply pytest-selenium/pages/base.py]: it will be some time before I can go through and add all the type hints.

File: pytest-selenium/pages/content.py
Comment (m1yag1) [reply pytest-selenium/pages/content.py]: Did you test this change?

File: pytest-selenium/tests/conftest.py
Comment (m1yag1) [reply pytest-selenium/tests/conftest.py]: Selenium uses these backwards from the way we typically see them. I've adjusted them in the documentation so that they use the typical order.

File: pytest-selenium/setup.cfg
Comment (m1yag1) [reply pytest-selenium/setup.cfg]: These are defaults passed into the pytest command. You run pytest --driver Chrome pytest-selenium/tests and these are inserted in there so you don't have to type them out.

File: pytest-selenium/tests/conftest.py
Comment (m1yag1) [reply pytest-selenium/tests/conftest.py]: Sure, it might but I'd rather not make changes on what might happen. This is basically copied from our other automation framework that does run in travis.

File: pytest-selenium/pages/content.py
Comment (philschatz) [reply pytest-selenium/pages/content.py]: eep, there are 36 `<nav>` elements. I think that's a bug... there should only be one.

File: pytest-selenium/setup.cfg
Comment (philschatz) [reply pytest-selenium/setup.cfg]: ah ok. I was just curious about the purpose of `sensitive_url` but thanks for enumerating the others

File: pytest-selenium/pages/content.py
Comment (philschatz) [reply pytest-selenium/pages/content.py]: This will probably be fixed to be just one `<nav>` as part of the "Add ToC Toggle" story in "Reading Experience": https://docs.google.com/document/d/1-KX5NMSeJTP5309L7mlXw85YHG_7XRFm-C7ofIdWtCE/edit#heading=h.apfasqkrfa3g

Comment (TomWoodward) [reply pytest-selenium/pages/content.py]: @philschatz the nav element is used for navs and subnavs, so each chapter has its own nav. i don't know of any reason to limit to just 1 nav per page

File: pytest-selenium/tests/conftest.py
Comment (m1yag1) [reply pytest-selenium/tests/conftest.py]: Actually it might be that pytest injects them in the opposite order.

File: pytest-selenium/pages/content.py
Comment (philschatz) [reply pytest-selenium/pages/content.py]: From the [MDN Description of nav](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/nav) it seems like they are used to denote separate navigations on a page (e.g. ToC, site navigation) and are not nested. In the epub spec the entire ToC is a `<nav>` which contains nested `<ul>`, `<li>`, and `<a href="">` elements.

Comment (TomWoodward) [reply pytest-selenium/pages/content.py]: ok, we can definitely change it to only be one nav for the ToC, we may end up having another one for site nav though so we should assume there are at least a couple

Comment (philschatz) [reply pytest-selenium/pages/content.py]: Yeppers, and when a 2nd `<nav>` is introduced they can be differentiated via `aria-label` attributes so the selenium tests _and_ screenreaders can differentiate 😄

**Review by philschatz — DISMISSED** [review]

**Comment by philschatz** [comment]
@m1yag1 Do you want to make any comments before this gets merged? I pushed a few commits addressing my review comments yesterday but forgot to ask you to review

**Review by philschatz — DISMISSED** [review]

File: README.md
Comment (TomWoodward) [inline README.md]: this command doesn't seem to exist

**Review by TomWoodward — DISMISSED** [review]

**Comment by m1yag1** [comment]
I did a merge of Phil’s code but had a conflict. I thought i put that in there. ugh. i think i messed something up.

**Review by TomWoodward — DISMISSED** [review]

File: pytest-selenium/pages/content.py
Comment (philschatz) [reply pytest-selenium/pages/content.py]: Even when the locator matches `> 1` elements? I'm hoping that selenium can:

- help ensure there is exactly 1 `h1` on the page
- explicitly fail when more than 1 element matches a selector

**Review by TomWoodward — DISMISSED** [review]

**Review by philschatz — DISMISSED** [review]
I just fixed a link and clarified that `make test` does not work by printing out the help & erroring so people know to run `make test-local` or `make test-rail`

File: pytest-selenium/pages/content.py
Comment (m1yag1) [reply pytest-selenium/pages/content.py]: We would need to write a test for that. Selenium doesn't enforce any standards like that.

Comment (m1yag1) [reply pytest-selenium/pages/content.py]: What could be automated is some sort of HTML validation. I don't know any tools that do that off the top of my head but i'm sure something could be integrating. It would be something very similar to an aXe report. We would have it fail when there were any failures or some threshold.

Comment (m1yag1) [reply pytest-selenium/pages/content.py]: Something like this seems like it would do what you want. https://validator.w3.org/docs/api.html

**Review by TomWoodward — APPROVED** [review]

**Review by philschatz — APPROVED** [review]

**Comment by philschatz** [comment]
oooh, pyyaml has a security vulnerability & should be upgraded to `pyyaml>=4.2b1`. https://github.com/openstax/rex-web/network/alerts

File: pytest-selenium/pages/content.py
Comment (philschatz) [reply pytest-selenium/pages/content.py]: (just bike-shedding)

Oh, I guess that I was trying to say that if the test-writer expects the selector to match one element (by using `find_element` rather than `find_elements`) then it seems like that assumption should throw an error if more than one element is matched. Otherwise, `find_element` really means `find_first_element`.

Once I contribute a test that uses `find_element` I might refactor all of the `find_element` references to be something like `util.find_element(self, ...)` or something.

**Comment by m1yag1** [comment]
Find element returns the first one it finds. If you want more than one you have to use find_elements. I do see where it would be useful.
  

  
>

**Pull request merged**
